cedar-policy-validator/src/custom_schema/ast.rs

```diff
@@ -0,0 +1,187 @@
+use cedar_policy_core::{
+    ast::{Id, Name as CName},
+    parser::{Loc, Node},
+};
+use itertools::Either;
+use miette::SourceSpan;
+use nonempty::NonEmpty;
+use smol_str::SmolStr;
+
+use super::to_json_schema::PrimitiveType;
+
+const IPADDR_EXTENSION: &str = "ipaddr";
+const DECIMAL_EXTENSION: &str = "decimal";
+pub(super) const CEDAR_NAMESPACE: &str = "__cedar";
+
+pub type Ident = Node<Id>;
+pub type Str = Node<SmolStr>;
+pub type Name = Either<Ident, Str>;
+pub type Schema = Vec<Node<Namespace>>;
+
+#[derive(Debug, Clone)]
+pub struct Path {
+    pub base: Ident,
+    pub prefix: Vec<Ident>,
+    pub(super) kind: PathKind,
+}
+
+#[derive(Debug, Clone)]
+pub(super) enum PathKind {
+    CedarBuiltin(Option<PrimOrExtension>),
+    Unqualified(Option<PrimOrExtension>),
+    Other,
+}
+
+#[derive(Debug, Clone)]
+pub(super) enum PrimOrExtension {
+    IpAddr,
+    Decimal,
+    Prim(PrimitiveType),
+}
+
+impl From<Ident> for Path {
+    fn from(base: Ident) -> Self {
+        let prefix = vec![];
+        let kind = Self::path_kind(&prefix, &base);
+        Path { base, prefix, kind }
+    }
+}
+
+impl Path {
+    pub fn new(base: Ident, prefix: impl IntoIterator<Item = Node<Id>>) -> Self {
+        let prefix = prefix.into_iter().collect::<Vec<_>>();
+        let kind = Self::path_kind(&prefix, &base);
+        Path { base, prefix, kind }
+    }
+
+    fn path_kind(prefix: &[Ident], base: &Ident) -> PathKind {
+        if Self::is_builtin(prefix) {
+            PathKind::CedarBuiltin(Self::parse_prim_or_ext(base))
+        } else if Self::is_unqualified(prefix) {
+            PathKind::Unqualified(Self::parse_prim_or_ext(base))
+        } else {
+            PathKind::Other
+        }
+    }
+
+    fn parse_prim_or_ext(base: &Ident) -> Option<PrimOrExtension> {
+        match base.node.as_ref() {
+            IPADDR_EXTENSION => Some(PrimOrExtension::IpAddr),
+            DECIMAL_EXTENSION => Some(PrimOrExtension::Decimal),
+            _ => Self::parse_prim_type(base).map(PrimOrExtension::Prim),
+        }
+    }
```

Comment:
not actually related to this code, but this made me think of it: We should have thorough tests for "Rules 1-5" in the RFC text.  For instance, that primitive vs entity name collisions are resolved in favor of entity, etc.  Perhaps we already have this, I'm just making a note here as I continue my review